File: src/main/team/crypto.ts
import { randomBytes, scryptSync, createCipheriv, createDecipheriv } from 'crypto'
import { FerroError } from '@shared/errors'
import {
  INVITE_CODE_PREFIX,
  type TeamBlobFile,
  type TeamInviteFile,
  type TeamInvitePayload,
  type TeamPayload,
  type TeamRole
} from '@shared/team'

// ── Ekip kripto katmanı ─────────────────────────────────────────────────────
// İki ayrı şifreleme sınırı:
//  1) Kasa: TeamPayload, EKİP ANAHTARI (TK) ile AES-256-GCM'le şifrelenir. TK
//     paroladan türetilmez — 32 baytlık tam entropili rastgele anahtardır, bu
//     yüzden KDF yoktur. (Sync tarafındaki parola-türetmeli şemadan bilinçli
//     olarak farklıdır: ekip anahtarı insanlar arasında paylaşılır, bir insan
//     parolası değildir.)
//  2) Davet: TeamInvitePayload (TK + depo jetonları), davet-PIN'inden scrypt ile
//     türetilen anahtarla şifrelenir. PIN kısa/insan-taşınabilir olabildiği için
//     KDF vardır (kaba kuvvet maliyetini yükseltir).
// Her iki tarafta da yanlış anahtar/PIN GCM etiket doğrulamasında yakalanır —
// düz metin asla kısmen açılmaz.

const IV_LEN = 12
const KEY_LEN = 32
const SALT_LEN = 16
const TAG_LEN = 16

/** Davet PIN'i için scrypt parametreleri (sync tarafıyla aynı; etkileşimli). */
const KDF = { algo: 'scrypt' as const, N: 16384, r: 8, p: 1 }

// ── Ekip anahtarı üretimi ──

/** Yeni bir ekip için tam entropili 32 baytlık anahtar üretir (base64). */
export function generateTeamKey(): string {
  return randomBytes(KEY_LEN).toString('base64')
}

function keyFromBase64(b64: string): Buffer {
  const key = Buffer.from(b64, 'base64')
  if (key.length !== KEY_LEN) {
    throw new FerroError('VALIDATION', 'Ekip anahtarı geçersiz (32 bayt olmalı)')
  }
  return key
}

// ── Düşük seviye AES-256-GCM yardımcıları (iv|ct|tag → base64 alanları) ──

function gcmEncrypt(key: Buffer, plain: Buffer): { iv: string; data: string } {
  const iv = randomBytes(IV_LEN)
  const cipher = createCipheriv('aes-256-gcm', key, iv)
  const enc = Buffer.concat([cipher.update(plain), cipher.final(), cipher.getAuthTag()])
  return { iv: iv.toString('base64'), data: enc.toString('base64') }
}

function gcmDecrypt(key: Buffer, ivB64: string, dataB64: string): Buffer {
  const raw = Buffer.from(dataB64, 'base64')
  if (raw.length < TAG_LEN + 1) throw new FerroError('VALIDATION', 'Şifreli veri eksik/bozuk')
  const tag = raw.subarray(raw.length - TAG_LEN)
  const body = raw.subarray(0, raw.length - TAG_LEN)
  const decipher = createDecipheriv('aes-256-gcm', key, Buffer.from(ivB64, 'base64'))
  decipher.setAuthTag(tag)
  return Buffer.concat([decipher.update(body), decipher.final()])
}

// ── Kasa şifrelemesi (ekip anahtarıyla) ──

/** TeamPayload'ı ekip anahtarıyla şifreleyip uzağa yazılacak zarfı üretir. */
export function encryptTeamPayload(payload: TeamPayload, teamKeyB64: string): TeamBlobFile {
  const key = keyFromBase64(teamKeyB64)
  const plain = Buffer.from(JSON.stringify(payload), 'utf8')
  const { iv, data } = gcmEncrypt(key, plain)
  return {
    app: 'ferro',
    kind: 'ferro-team',
    version: 1,
    teamId: payload.teamId,
    name: payload.name,
    updatedAt: payload.updatedAt,
    revision: payload.revision,
    cipher: 'aes-256-gcm',
    iv,
    data
  }
}

/**
 * Uzaktan inen kasa zarfını çözer. Yanlış ekip anahtarı / bozuk veri
 * AUTH_FAILED (GCM etiketi tutmaz); tanınmayan format VALIDATION fırlatır.
 */
export function decryptTeamPayload(file: TeamBlobFile, teamKeyB64: string): TeamPayload {
  if (file.kind !== 'ferro-team' || file.version !== 1 || file.cipher !== 'aes-256-gcm') {
    throw new FerroError('VALIDATION', 'Tanınmayan ekip kasası formatı')
  }
  const key = keyFromBase64(teamKeyB64)
  let plain: Buffer
  try {
    plain = gcmDecrypt(key, file.iv, file.data)
  } catch {
    throw new FerroError('AUTH_FAILED', 'Ekip anahtarı hatalı ya da kasa bozulmuş')
  }
  try {
    const parsed = JSON.parse(plain.toString('utf8')) as TeamPayload
    if (parsed.kind !== 'ferro-team-payload') throw new Error('kind uyuşmuyor')
    return parsed
  } catch (err) {
    throw new FerroError('VALIDATION', 'Ekip kasası yükü çözümlenemedi', String(err))
  }
}

// ── Davet şifrelemesi (PIN ile) ──

function deriveInviteKey(pin: string, salt: Buffer, N: number, r: number, p: number): Buffer {
  return scryptSync(pin, salt, KEY_LEN, { N, r, p, maxmem: 128 * 1024 * 1024 })
}

/** TeamInvitePayload'ı davet-PIN'iyle şifreleyip davet zarfını üretir. */
export function encryptInvite(
  payload: TeamInvitePayload,
  pin: string,
  meta: { teamId: string; teamName: string; role: TeamRole }
): TeamInviteFile {
  const salt = randomBytes(SALT_LEN)
  const key = deriveInviteKey(pin, salt, KDF.N, KDF.r, KDF.p)
  const { iv, data } = gcmEncrypt(key, Buffer.from(JSON.stringify(payload), 'utf8'))
  return {
    app: 'ferro',
    kind: 'ferro-invite',
    version: 1,
    teamId: meta.teamId,
    teamName: meta.teamName,
    role: meta.role,
    kdf: { algo: KDF.algo, salt: salt.toString('base64'), N: KDF.N, r: KDF.r, p: KDF.p },
    cipher: 'aes-256-gcm',
    iv,
    data
  }
}

/** Davet zarfını PIN ile çözer. Yanlış PIN → AUTH_FAILED. */
export function decryptInvite(file: TeamInviteFile, pin: string): TeamInvitePayload {
  if (file.kind !== 'ferro-invite' || file.version !== 1 || file.cipher !== 'aes-256-gcm') {
    throw new FerroError('VALIDATION', 'Tanınmayan davet formatı')
  }
  const { salt, N, r, p } = file.kdf
  // KDF parametreleri davet kodundan gelir — güvensiz girdi; maliyet sınırlanır.
  if (N > 1 << 20 || r > 16 || p > 4) {
    throw new FerroError('VALIDATION', 'Davetin KDF parametreleri desteklenmiyor')
  }
  const key = deriveInviteKey(pin, Buffer.from(salt, 'base64'), N, r, p)
  let plain: Buffer
  try {
    plain = gcmDecrypt(key, file.iv, file.data)
  } catch {
    throw new FerroError('AUTH_FAILED', 'Davet PIN’i hatalı ya da davet kodu bozuk')
  }
  try {
    return JSON.parse(plain.toString('utf8')) as TeamInvitePayload
  } catch (err) {
    throw new FerroError('VALIDATION', 'Davet yükü çözümlenemedi', String(err))
  }
}

// ── Davet kodu kodlama/çözme (zarf ↔ taşınabilir string) ──

/** Davet zarfını tek satırlık taşınabilir koda çevirir. */
export function encodeInviteCode(file: TeamInviteFile): string {
  const json = JSON.stringify(file)
  return INVITE_CODE_PREFIX + Buffer.from(json, 'utf8').toString('base64url')
}

/** Davet kodunu zarfa geri çözer. Biçim tanınmazsa VALIDATION fırlatır. */
export function decodeInviteCode(code: string): TeamInviteFile {
  const trimmed = code.trim()
  if (!trimmed.startsWith(INVITE_CODE_PREFIX)) {
    throw new FerroError('VALIDATION', 'Geçersiz davet kodu (önek uyuşmuyor)')
  }
  const b64 = trimmed.slice(INVITE_CODE_PREFIX.length)
  let file: TeamInviteFile
  try {
    file = JSON.parse(Buffer.from(b64, 'base64url').toString('utf8')) as TeamInviteFile
  } catch (err) {
    throw new FerroError('VALIDATION', 'Davet kodu çözümlenemedi', String(err))
  }
  if (file.app !== 'ferro' || file.kind !== 'ferro-invite') {
    throw new FerroError('VALIDATION', 'Bu bir Ferro davet kodu değil')
  }
  return file
}

File: src-tauri/src/team/crypto.rs
//! Ekip kasası + davet şifrelemesi — `team/crypto.ts` karşılığı (byte-uyumlu).
//!
//! İki sınır: (1) Kasa, tam entropili 32B ekip anahtarıyla (TK) şifrelenir — KDF yok.
//! (2) Davet, PIN'den scrypt ile türetilen anahtarla şifrelenir. Her ikisi de
//! AES-256-GCM, `data = base64(ct ‖ tag)`, `iv` ayrı.

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use serde::{Deserialize, Serialize};

use crate::crypto::{
    b64, derive_key, gcm_decrypt, gcm_encrypt, key_from_b64, random_bytes, unb64, KDF_N, KDF_P,
    KDF_R, SALT_LEN,
};
use crate::error::{FerroError, FerroErrorCode, FerroResult};
use crate::types::SiteExportEntry;

pub const TEAM_FILE_NAME: &str = "ferro-team.json";
pub const INVITE_CODE_PREFIX: &str = "ferro-invite-1.";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamMember {
    pub id: String,
    pub name: String,
    pub role: String, // admin | member | readonly
    #[serde(rename = "addedAt")]
    pub added_at: String,
    #[serde(rename = "addedBy", default, skip_serializing_if = "Option::is_none")]
    pub added_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamPayload {
    pub kind: String, // 'ferro-team-payload'
    pub version: u32,
    #[serde(rename = "teamId")]
    pub team_id: String,
    pub name: String,
    #[serde(rename = "updatedAt")]
    pub updated_at: String,
    pub revision: u64,
    pub members: Vec<TeamMember>,
    pub sites: Vec<SiteExportEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamBlobFile {
    pub app: String,
    pub kind: String,
    pub version: u32,
    #[serde(rename = "teamId")]
    pub team_id: String,
    pub name: String,
    #[serde(rename = "updatedAt")]
    pub updated_at: String,
    pub revision: u64,
    pub cipher: String,
    pub iv: String,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InviteGist {
    #[serde(rename = "gistId")]
    pub gist_id: String,
    pub token: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InviteWebdav {
    pub url: String,
    pub user: String,
    pub password: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamInvitePayload {
    #[serde(rename = "teamKey")]
    pub team_key: String,
    pub provider: String, // gist | webdav
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gist: Option<InviteGist>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub webdav: Option<InviteWebdav>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InviteKdf {
    pub algo: String,
    pub salt: String,
    #[serde(rename = "N")]
    pub n: u32,
    pub r: u32,
    pub p: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamInviteFile {
    pub app: String,
    pub kind: String,
    pub version: u32,
    #[serde(rename = "teamId")]
    pub team_id: String,
    #[serde(rename = "teamName")]
    pub team_name: String,
    pub role: String,
    pub kdf: InviteKdf,
    pub cipher: String,
    pub iv: String,
    pub data: String,
}

/// Yeni bir ekip anahtarı üretir (base64, 32 bayt tam entropi).
pub fn generate_team_key() -> String {
    b64(&random_bytes(32))
}

// ── Kasa (TK ile) ──

pub fn encrypt_team_payload(
    payload: &TeamPayload,
    team_key_b64: &str,
) -> FerroResult<TeamBlobFile> {
    let key = key_from_b64(team_key_b64)?;
    let plain = serde_json::to_vec(payload)?;
    let (iv, data) = gcm_encrypt(&key, &plain)?;
    Ok(TeamBlobFile {
        app: "ferro".into(),
        kind: "ferro-team".into(),
        version: 1,
        team_id: payload.team_id.clone(),
        name: payload.name.clone(),
        updated_at: payload.updated_at.clone(),
        revision: payload.revision,
        cipher: "aes-256-gcm".into(),
        iv: b64(&iv),
        data: b64(&data),
    })
}

pub fn decrypt_team_payload(file: &TeamBlobFile, team_key_b64: &str) -> FerroResult<TeamPayload> {
    if file.kind != "ferro-team" || file.version != 1 || file.cipher != "aes-256-gcm" {
        return Err(FerroError::new(
            FerroErrorCode::Validation,
            "Tanınmayan ekip dosyası formatı",
        ));
    }
    let key = key_from_b64(team_key_b64)?;
    let plain = gcm_decrypt(&key, &unb64(&file.iv)?, &unb64(&file.data)?)?;
    let payload: TeamPayload = serde_json::from_slice(&plain).map_err(|e| {
        FerroError::with_detail(
            FerroErrorCode::Validation,
            "Ekip yükü çözümlenemedi",
            e.to_string(),
        )
    })?;
    if payload.kind != "ferro-team-payload" {
        return Err(FerroError::new(
            FerroErrorCode::Validation,
            "Ekip yükü türü uyuşmuyor",
        ));
    }
    Ok(payload)
}

// ── Davet (PIN ile) ──

pub fn encrypt_invite(
    payload: &TeamInvitePayload,
    pin: &str,
    team_id: &str,
    team_name: &str,
    role: &str,
) -> FerroResult<TeamInviteFile> {
    let salt = random_bytes(SALT_LEN);
    let key = derive_key(pin, &salt, KDF_N, KDF_R, KDF_P)?;
    let plain = serde_json::to_vec(payload)?;
    let (iv, data) = gcm_encrypt(&key, &plain)?;
    Ok(TeamInviteFile {
        app: "ferro".into(),
        kind: "ferro-invite".into(),
        version: 1,
        team_id: team_id.into(),
        team_name: team_name.into(),
        role: role.into(),
        kdf: InviteKdf {
            algo: "scrypt".into(),
            salt: b64(&salt),
            n: KDF_N,
            r: KDF_R,
            p: KDF_P,
        },
        cipher: "aes-256-gcm".into(),
        iv: b64(&iv),
        data: b64(&data),
    })
}

pub fn decrypt_invite(file: &TeamInviteFile, pin: &str) -> FerroResult<TeamInvitePayload> {
    if file.kind != "ferro-invite" || file.version != 1 || file.cipher != "aes-256-gcm" {
        return Err(FerroError::new(
            FerroErrorCode::Validation,
            "Tanınmayan davet formatı",
        ));
    }
    let salt = unb64(&file.kdf.salt)?;
    let key = derive_key(pin, &salt, file.kdf.n, file.kdf.r, file.kdf.p)?;
    let plain = gcm_decrypt(&key, &unb64(&file.iv)?, &unb64(&file.data)?)?;
    serde_json::from_slice(&plain).map_err(|e| {
        FerroError::with_detail(
            FerroErrorCode::Validation,
            "Davet yükü çözümlenemedi",
            e.to_string(),
        )
    })
}

// ── Davet kodu (taşınabilir string) ──

pub fn encode_invite_code(file: &TeamInviteFile) -> FerroResult<String> {
    let json = serde_json::to_vec(file)?;
    Ok(format!(
        "{INVITE_CODE_PREFIX}{}",
        URL_SAFE_NO_PAD.encode(json)
    ))
}

pub fn decode_invite_code(code: &str) -> FerroResult<TeamInviteFile> {
    let rest = code
        .strip_prefix(INVITE_CODE_PREFIX)
        .ok_or_else(|| FerroError::new(FerroErrorCode::Validation, "Geçersiz davet kodu öneki"))?;
    let bytes = URL_SAFE_NO_PAD
        .decode(rest.trim())
        .map_err(|_| FerroError::new(FerroErrorCode::Validation, "Davet kodu çözülemedi"))?;
    let file: TeamInviteFile = serde_json::from_slice(&bytes).map_err(|e| {
        FerroError::with_detail(
            FerroErrorCode::Validation,
            "Davet kodu biçimi geçersiz",
            e.to_string(),
        )
    })?;
    if file.app != "ferro" || file.kind != "ferro-invite" {
        return Err(FerroError::new(
            FerroErrorCode::Validation,
            "Tanınmayan davet kodu",
        ));
    }
    Ok(file)
}
